src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -108,6 +122,94 @@ ShouldCreateTenantSchemaTable(Oid relationId)
 }
 
 
+/*
+ * EnsureTableKindSupportedForTenantSchema ensures that given table's kind is
+ * supported by a tenant schema.
+ */
+static void
+EnsureTableKindSupportedForTenantSchema(Oid relationId)
+{
+	if (IsForeignTable(relationId))
+	{
+		ereport(ERROR, (errmsg("cannot create a foreign table in a distributed "
+							   "schema")));
+	}
+	else if (PartitionTable(relationId))
+	{
+		ErrorIfIllegalPartitioningInTenantSchema(PartitionParentOid(relationId),
+												 relationId);
+	}
+	else if (IsChildTable(relationId) || IsParentTable(relationId))
+	{
+		ereport(ERROR, (errmsg("tables in a distributed schema cannot inherit or "
+							   "be inherited")));
+	}
+}
+
+
+/*
+ * EnsureFKeysForTenantTable ensures that all referencing and referenced foreign
+ * keys are allowed for given table.
+ */
+static void
+EnsureFKeysForTenantTable(Oid relationId)
+{
+	Oid tenantSchemaId = get_rel_namespace(relationId);
+	int fKeyReferencingFlags = INCLUDE_REFERENCING_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
```

Comment:
Can use EXCLUDE_SELF_REFERENCES so that we don't have to deal with self-referencing fkeys in below loop?